File: test/python/misc.py
# coding=utf-8
#
#     This file is part of CasADi.
#
#     CasADi -- A symbolic framework for dynamic optimization.
#     Copyright (C) 2010-2023 Joel Andersson, Joris Gillis, Moritz Diehl,
#                             KU Leuven. All rights reserved.
#     Copyright (C) 2011-2014 Greg Horn
#
#     CasADi is free software; you can redistribute it and/or
#     modify it under the terms of the GNU Lesser General Public
#     License as published by the Free Software Foundation; either
#     version 3 of the License, or (at your option) any later version.
#
#     CasADi is distributed in the hope that it will be useful,
#     but WITHOUT ANY WARRANTY; without even the implied warranty of
#     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
#     Lesser General Public License for more details.
#
#     You should have received a copy of the GNU Lesser General Public
#     License along with CasADi; if not, write to the Free Software
#     Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
#
#
import casadi as ca
import casadi as c
import numpy
import unittest
from types import *
from helpers import *
import pickle
from operator import itemgetter
import sys
from casadi.tools import capture_stdout

scipy_available = True
try:
    import scipy.special
    from scipy.linalg import expm
except:
    scipy_available = False

class Misctests(casadiTestCase):

  def test_issue179B(self):
    self.message('Regression test #179 (B)')
    def calc_sparsity():
      x = ca.SX.sym("x")
      f = ca.Function('f', [x], [x ** 2])
      return f.jac_sparsity(0, 0)

    def print_sparsity():
        sparsity = calc_sparsity()
        str(sparsity) # Segfault

    print_sparsity()

  def test_issue4216(self):
    self.message('Regression test #4216: complex numpy inputs should raise, not segfault')
    self.assertRaises(TypeError, lambda: 0.5j * ca.MX(1))  # pyright: ignore[reportCallIssue,reportArgumentType,reportOperatorIssue,reportAttributeAccessIssue]

    c = numpy.array([3.+2j])
    self.assertRaises((TypeError, NotImplementedError), lambda: ca.DM(c))  # pyright: ignore[reportCallIssue,reportArgumentType,reportOperatorIssue,reportAttributeAccessIssue]
    self.assertRaises((TypeError, NotImplementedError), lambda: ca.SX(c))  # pyright: ignore[reportCallIssue,reportArgumentType,reportOperatorIssue,reportAttributeAccessIssue]
    self.assertRaises((TypeError, NotImplementedError), lambda: ca.MX(c))  # pyright: ignore[reportCallIssue,reportArgumentType,reportOperatorIssue,reportAttributeAccessIssue]

    cv = 0.5j * numpy.ones(2)
    # MX has no implicit numeric conversion so matmul reaches the casadi
    # path and must raise TypeError. For DM/SX, numpy's __array__ path
    # may instead raise ValueError on shape mismatch; either is fine here
    # as long as no segfault occurs.
    self.assertRaises((TypeError, NotImplementedError), lambda: cv @  ca.MX(2))  # pyright: ignore[reportCallIssue,reportArgumentType,reportOperatorIssue,reportAttributeAccessIssue]
    for d in (ca.DM(2), ca.SX(2)):
      self.assertRaises((TypeError, NotImplementedError, ValueError),
                        lambda d=d: cv @ d)  # pyright: ignore[reportCallIssue,reportArgumentType,reportOperatorIssue,reportAttributeAccessIssue]

  def test_sanity(self):
    ca.DM(ca.Sparsity(4,3,[0,2,2,3],[1,2,1]),[0.738,0.39,0.99])
    self.assertRaises(RuntimeError,lambda : ca.DM(ca.Sparsity(4,4,[0,2,2,3],[1,2,1]),[0.738,0.39,0.99]))
    self.assertRaises(RuntimeError,lambda : ca.DM(ca.Sparsity(4,3,[0,2,2,12],[1,2,1]),[0.738,0.39,0.99]))
    self.assertRaises(RuntimeError,lambda : ca.DM(ca.Sparsity(4,3,[-10,2,2,3],[1,2,1]),[0.738,0.39,0.99]))
    self.assertRaises(RuntimeError,lambda : ca.DM(ca.Sparsity(4,3,[0,2,2,3],[8,2,1]),[0.738,0.39,0.99]))
    self.assertRaises(RuntimeError,lambda : ca.DM(ca.Sparsity(4,3,[0,2,2,3],[-3,2,1]),[0.738,0.39,0.99]))
    self.assertRaises(RuntimeError,lambda : ca.DM(ca.Sparsity(4,3,[0,2,2,3],[1,2,1,2]),[0.738,0.39,0.99]))
    self.assertRaises(RuntimeError,lambda : ca.DM(ca.Sparsity(4,3,[0,2,0,3],[1,2,1]),[0.738,0.39,0.99]))

  def test_copyconstr_norefcount(self):
    self.message("Copy constructor for non-refcounted classes")
    x = ca.DM.ones(2,3)

    y = ca.DM(x)
    x[0,0] = 5

    self.assertFalse(id(x)==id(y))
    self.assertEqual(x[0,0],5)
    self.assertEqual(y[0,0],1)

  def test_copyconstr_refcount(self):
    self.message("Copy constructor for refcounted classes")
    x = ca.Sparsity.diag(4)

    y = ca.Sparsity(x)

    x.resize(2,8)

    self.assertFalse(id(x)==id(y))
    self.assertTrue(x.numel(),y.numel())
    self.checkarray(x.shape,(2,8),"shape")
    self.checkarray(y.shape,(4,4),"shape")

  def test_copy_norefcount(self):
    self.message("Shallow copy for non-refcounted classes")
    import copy

    x = ca.DM.ones(2,3)

    y = copy.copy(x)
    x[0,0] = 5

    self.assertFalse(id(x)==id(y))
    self.assertEqual(x[0,0],5)
    self.assertEqual(y[0,0],1)

  def test_copy_refcount(self):
    self.message("Shallow copy for refcounted classes")
    import copy
    x = ca.Sparsity.diag(4)

    y = copy.copy(x)

    x.resize(2,8)

    self.assertFalse(id(x)==id(y))
    self.assertTrue(x.numel(),y.numel())
    self.checkarray(x.shape,(2,8),"shape")
    self.checkarray(y.shape,(4,4),"shape")

  def test_deepcopy_norefcount(self):
    self.message("Deep copy for non-refcounted classes")
    import copy

    x = ca.DM.ones(2,3)

    y = copy.deepcopy(x)
    x[0,0] = 5

    self.assertFalse(id(x)==id(y))
    self.assertEqual(x[0,0],5)
    self.assertEqual(y[0,0],1)

  def test_deepcopy_refcount(self):
    self.message("Deep copy for refcounted classes")
    import copy
    x = ca.Sparsity.diag(4)

    y = copy.deepcopy(x)

    x.resize(2,8)

    self.assertFalse(id(x)==id(y))
    self.assertTrue(x.numel(),y.numel())
    self.checkarray(x.shape,(2,8),"shape")
    self.checkarray(y.shape,(4,4),"shape")

  @requiresPlugin(ca.nlpsol,"ipopt")
  def test_options_introspection(self):
    self.message("options introspection")
    x=ca.SX.sym("x")
    nlp = {'x':x, 'f':x**2}
    i = ca.nlpsol('i', "ipopt", nlp)

    opts = i.optionNames()  # pyright: ignore[reportAttributeAccessIssue]
    self.assertTrue(isinstance(opts,list))

    n = opts[0]
    self.assertTrue(type(n)==type(""))

    n = "monitor"

    d = i.optionDescription(n)  # pyright: ignore[reportAttributeAccessIssue]
    self.assertTrue(type(d)==type(""))
    self.assertTrue(not("d"=="N/A"))

    d = i.optionTypeName(n)  # pyright: ignore[reportAttributeAccessIssue]
    self.assertEqual(d,"OT_STRINGVECTOR")

    #d = i.optionAllowed(n)

  def test_pickling(self):

    a = ca.Sparsity.lower(4)
    s = pickle.dumps(a)
    b = pickle.loads(s)
    self.assertTrue(a==b)

    a = ca.Sparsity()
    s = pickle.dumps(a)
    b = pickle.loads(s)
    self.assertTrue(a.is_null())

    a = ca.DM(ca.Sparsity.lower(4),list(range(10)))
    s = pickle.dumps(a)
    b = pickle.loads(s)
    self.checkarray(a,b)

  def test_exceptions(self):
    if systemswig: return

    try:
      ca.nlpsol(123)  # pyright: ignore[reportCallIssue,reportArgumentType,reportOperatorIssue,reportAttributeAccessIssue]
      self.assertTrue(False)
    except NotImplementedError as e:
      e_message = str(e)
      assert "nlpsol(str,str,dict:MX,dict)" in e_message
      assert "You have: '(int)'" in e_message
      assert "::" not in e_message
      assert "std" not in e_message

    try:
      ca.vcat(123)  # pyright: ignore[reportCallIssue,reportArgumentType,reportOperatorIssue,reportAttributeAccessIssue]
      self.assertTrue(False)
    except NotImplementedError as e:
      e_message = str(e)
      assert "vertcat([SX]" in e_message
      assert "vertcat([DM]" in e_message
      assert "You have: '(int)'" in e_message
      assert "::" not in e_message
      assert "std" not in e_message

    try:
      ca.substitute(123)  # pyright: ignore[reportCallIssue,reportArgumentType,reportOperatorIssue,reportAttributeAccessIssue]
      self.assertTrue(False)
    except NotImplementedError as e:
      e_message = str(e)
      assert "substitute(SX,SX,SX)" in e_message
      assert "substitute([SX],[SX],[SX])" in e_message
      assert "You have: '(int)'" in e_message
      assert "::" not in e_message
      assert "std" not in e_message

    try:
      ca.load_nlpsol(132)  # pyright: ignore[reportCallIssue,reportArgumentType,reportOperatorIssue,reportAttributeAccessIssue]
      self.assertTrue(False)
    except NotImplementedError as e:
      e_message = str(e)
      assert "load_nlpsol(str)" in e_message
      assert "::" not in e_message
      assert "std" not in e_message

    x=ca.SX.sym("x")

    try:
      [x]+ x  # pyright: ignore[reportCallIssue,reportArgumentType,reportOperatorIssue,reportAttributeAccessIssue]
      self.assertTrue(False)
    except TypeError as e:
      e_message = str(e)

    try:
      x + [x]  # pyright: ignore[reportCallIssue,reportArgumentType,reportOperatorIssue,reportAttributeAccessIssue]
      self.assertTrue(False)
    except TypeError as e:
      e_message = str(e)

    try:
      x.reshape(2)
      self.assertTrue(False)
    except NotImplementedError as e:
      e_message = str(e)
      assert "reshape(SX,(int,int))" in e_message

    try:
      x.reshape(("a",2))  # pyright: ignore[reportCallIssue,reportArgumentType,reportOperatorIssue,reportAttributeAccessIssue]
      self.assertTrue(False)
    except NotImplementedError as e:
      e_message = str(e)
      assert "You have: '(SX,(str,int))'" in e_message

    try:
      ca.diagsplit("s")  # pyright: ignore[reportCallIssue,reportArgumentType,reportOperatorIssue,reportAttributeAccessIssue]
      self.assertTrue(False)
    except NotImplementedError as e:
      e_message = str(e)
      assert "diagsplit(SX,int)" in e_message
      assert "diagsplit(DM,int)" in e_message

    try:
      ca.DM("df")  # pyright: ignore[reportCallIssue,reportArgumentType,reportOperatorIssue,reportAttributeAccessIssue]
      self.assertTrue(False)
    except NotImplementedError as e:
      e_message = str(e)
      assert "  DM(" in e_message

    try:
      ca.vertcat(1,ca.SX.sym('x'),ca.MX.sym('x'))  # pyright: ignore[reportCallIssue,reportArgumentType,reportOperatorIssue,reportAttributeAccessIssue]
      self.assertTrue(False)
    except NotImplementedError as e:
      e_message = str(e)
      assert "vertcat(" in e_message

  def test_getscheme(self):
    x = ca.SX.sym("x")
    p = ca.SX.sym("p")

    F = ca.Function('F', [x, p], [x+p, x**2], ['x', 'p'], ['f', 'g'])

    fc = F(x=3,p=4)
    f = fc['f']
    self.checkarray(f,ca.DM([7]))
    g = fc['g']
    self.checkarray(g,ca.DM([9]))
    [f,g] = itemgetter('f','g')(fc)
    self.checkarray(f,ca.DM([7]))
    self.checkarray(g,ca.DM([9]))
    [g,f] = itemgetter('g','f')(fc)
    self.checkarray(f,ca.DM([7]))
    self.checkarray(g,ca.DM([9]))

  def test_assertions(self):

    x = ca.MX.sym("x")

    z = x**2

    z = z.attachAssert(z>3,"x must be larger than 3")

    v = ca.sin(z)

    f = ca.Function('f', [x],[v])

    print(f)

    f_out = f(-6)
    try :
      f_out = f(1)
    except Exception as e:
      print(str(e))
      self.assertTrue("x must be larger than 3" in str(e))

  def test_doc(self):
    doc = ca.nlpsol.__doc__ or ""
    self.assertTrue("jac_penalty" in doc) # FunctionInternal
    self.assertTrue("print_time" in doc)  # ProtoFunction
    self.assertTrue( doc.count("print_time")==1)  # ProtoFunction

  @requires_nlpsol("ipopt")
  def test_output(self):
    with capture_stdout() as result:
      ca.DM([1,2]).print_dense()


    assert "2" in result[0]

    x=ca.SX.sym("x")
    f = {'x':x, 'f':x**2}
    solver = ca.nlpsol("solver", "ipopt",f)
    with capture_stdout() as result:
      solver_out = solver(x0=0)

    assert "Number of nonzeros in equality constraint" in result[0]
    assert "iter    objective    inf_pr" in result[0]

    with capture_stdout() as result:
      try:
        solver = ca.nlpsol("solver","foo",f)
      except:
        pass

    assert "casadi_nlpsol_foo" in result[1]

  def test_serialize(self):

    x = ca.Sparsity.upper(3)

    y = ca.SX.sym("y") # nested

    si = ca.StringSerializer()
    si.pack([x])

    z = y
    for i in range(10000):
      z = ca.sin(z)
    e = ca.vertcat(y,z,2*z)
    fref = ca.Function('f',[y],[e])
    si.pack(e)
    data = si.encode()
    si.pack([x])

    si = ca.StringDeserializer(data)

    spx = si.unpack()
    self.assertTrue(spx[0]==x)
    A = si.unpack()
    f = ca.Function('f',[A[0]],[A])
    self.checkfunction_light(f,fref,[7])
    with self.assertInException("end of stream"):
      si.unpack()
    with self.assertInException("end of stream"):
      si.unpack()

    si = ca.FileSerializer("foo.dat")
    si.pack([x])
    z = ca.sin(y)
    e = ca.vertcat(y,z,2*z)
    fref = ca.Function('f',[y],[e])
    si.pack(e)
    si = None
    si = ca.FileDeserializer("foo.dat")

    spx = si.unpack()
    self.assertTrue(spx[0]==x)
    A = si.unpack()
    f = ca.Function('f',[A[0]],[A])
    self.checkfunction_light(f,fref,[7])
    with self.assertInException("end of stream"):
      si.unpack()
    with self.assertInException("end of stream"):
      si.unpack()


    x = ca.SX.sym('x')
    s = ca.StringSerializer()
    s.pack(x)
    data1 = s.encode()
    s.pack(ca.sin(x))
    data2 = s.encode()

    s = ca.StringDeserializer(data1)
    a = s.unpack()
    with self.assertInException("end of stream"):
      s.unpack()
    s.decode(data2)
    b = s.unpack()
    with self.assertInException("end of stream"):
      s.unpack()
    si = None

    si = ca.FileSerializer("foo.dat")
    f = ca.Function("f",[x],[x**2])
    g = ca.Function("g",[x],[x**3])
    si.pack([f,g])
    si = None
    with self.assertInException("File is not loadable with 'load'. Use 'FileDeserializer' instead."):
      r = ca.Function.load("foo.dat")

    si = ca.FileDeserializer("foo.dat")
    print(si.unpack())

    si = None

    si = ca.FileSerializer("foo.dat")
    f = ca.Function("f",[x],[x**2])
    si.pack(f)
    si = None
    r = ca.Function.load("foo.dat")
    print(r)

    f.save("foo.dat")
    si = ca.FileDeserializer("foo.dat")
    print(si.unpack())

  def test_print_time(self):


    x = ca.MX.sym("x")
    f = x**2


    for print_time in [True,False]:
      included = ["t_wall"] if print_time else []
      excluded = [] if print_time else ["t_wall"]
      opts = {"print_time":print_time}
      ff = ca.Function("f",[x],[f],opts)
      with self.assertOutput(included, excluded):
        ff(3)

      if ca.has_nlpsol("ipopt"):
        solver = ca.nlpsol("nlpsol","ipopt",{"x":x,"f":f},opts)
        with self.assertOutput(included, excluded):
          solver()


      if ca.has_conic("ipopt"):
        solver = ca.qpsol("qpsol","qpoases",{"x":x,"f":f},opts)
        with self.assertOutput(included, excluded):
          solver()

      solver = ca.rootfinder("rootfinder","newton",{"x":x,"g":x},opts)
      with self.assertOutput(included, excluded):
        solver()

      solver = ca.integrator("integrator","rk",{"x":x,"ode":f},opts)
      with self.assertOutput(included, excluded):
        solver()


      integr_options = dict(opts)
      integr_options["simplify"] = True
      solver = ca.integrator("integrator","rk",{"x":x,"ode":f},integr_options)
      with self.assertOutput(included, excluded):
        solver()

      A = ca.DM.rand(3,3)
      if ca.has_linsol("lapacklu"):
        with self.assertOutput(included, excluded):
          ca.solve(A,ca.vertcat(1,2,3),"lapacklu",opts)


      Amx = ca.MX(ca.DM.rand(3,3))
      if ca.has_linsol("lapacklu"):
        with self.assertOutput(included, excluded):
          ca.evalf(ca.solve(Amx,ca.vertcat(1,2,3),"lapacklu",opts))

      if ca.has_linsol("lapacklu"):
        solver = ca.Linsol("linsol","lapacklu",A.sparsity(),opts)
        with self.assertOutput(included, excluded):
          solver.solve(A,ca.vertcat(1,2,3))

  def test_error_formatting(self):
    x = ca.MX.sym("x")

    f = ca.Function('f',[x],[x.attachAssert(0,"Hey \\W %d")])

    with self.assertInException("Hey \\W %d"):
        f()

  @memory_heavy()
  def test_record_time(self):


    x = ca.MX.sym("x")
    f = x**2
    for i in range(10000):
      f =ca.sin(f)*f
    f=f+x**2

    opts = {"record_time":True}
    ff = ca.Function("f",[x],[f],opts)
    ff(3)
    self.assertTrue("t_proc_total" in ff.stats())
    self.assertTrue(ff.stats()["t_wall_total"]>=0)

    if ca.has_nlpsol("ipopt"):
      solver = ca.nlpsol("nlpsol","ipopt",{"x":x,"f":f},opts)
      solver()
      self.assertTrue("t_proc_total" in solver.stats())
      self.assertTrue(solver.stats()["t_wall_total"]>0)

    if ca.has_conic("qpoases"):
      solver = ca.qpsol("qpsol","qpoases",{"x":x,"f":f},opts)
      solver()
      self.assertTrue("t_proc_total" in solver.stats())
      self.assertTrue(solver.stats()["t_wall_total"]>0)

    solver = ca.rootfinder("rootfinder","newton",{"x":x,"g":x},opts)
    solver()
    self.assertTrue("t_proc_total" in solver.stats())
    self.assertTrue(solver.stats()["t_wall_total"]>=0)

    solver = ca.integrator("integrator","rk",{"x":x,"ode":f},opts)
    solver()
    self.assertTrue("t_proc_total" in solver.stats())
    self.assertTrue(solver.stats()["t_wall_total"]>=0)

    integr_options = {}
    integr_options["simplify"] = True
    integr_options["simplify_options"] = opts
    solver = ca.integrator("integrator","rk",{"x":x,"ode":f},integr_options)
    solver()
    self.assertTrue("t_proc_total" in solver.stats())
    self.assertTrue(solver.stats()["t_wall_total"]>=0)

    A = ca.DM.rand(3,3)
    if ca.has_linsol("lapacklu"):
      solver = ca.Linsol("linsol","lapacklu",A.sparsity(),opts)
      solver.solve(A,ca.vertcat(1,2,3))
      self.assertTrue("t_proc_total" in solver.stats())
      self.assertTrue(solver.stats()["t_wall_total"]>=0)

  def test_unicode(self):
    import sys
    import shutil
    if sys.version_info[0] < 3: return
    path = "üni/tüst.casadi"
    path2 = "üni/stüst.casadi"
    if "ghc-filesystem" not in ca.CasadiMeta.feature_list():
        path = "tüst.casadi"
        path2 = "stüst.casadi"
    x = ca.MX.sym("x")
    f = ca.Function("f",[x],[x**2])
    f.save(path)
    shutil.copy(path,path2)
    
    f2 = ca.Function.load(path2)
    self.checkfunction_light(f,f2,inputs=[3])
    
    
if __name__ == '__main__':
    unittest.main()

File: test/javascript/misc.js
// misc.js -- port of test/python/misc.py (5/21 tests).
//
// Skipped (rationale at the file footer):
//   test_issue4216, test_options_introspection, test_pickling,
//   test_exceptions, test_doc, test_output, test_serialize,
//   test_print_time, test_error_formatting, test_record_time,
//   test_unicode, test_copyconstr_refcount, test_copy_norefcount,
//   test_copy_refcount, test_deepcopy_*

const { assertEqual, assertTrue, assertThrows, checkArray, runTests } = require("./_helpers");

function test_issue179B(M) {
  // Regression #179: jac_sparsity then str on the result used to segfault.
  const x = M.SX.sym("x");
  const f = M.Function("f", [x], [M.times(x, x)]);
  // jac_sparsity(oind, iind, symmetric) -- 3-arg overload in wasm-js.
  const s = f.jac_sparsity(0n, 0n, false);
  // Just verify str(sparsity) doesn't crash.
  const repr = "" + s;
  assertTrue(typeof repr === "string", "str(Sparsity) is string");
}

function test_sanity(M) {
  // DM constructed with invalid Sparsity arguments must throw, not segfault.
  // Valid baseline:
  const ok = new M.DM(new M.Sparsity(4n, 3n, [0n, 2n, 2n, 3n], [1n, 2n, 1n]),
                      [0.738, 0.39, 0.99]);
  assertTrue(ok != null, "valid DM constructs");
  // Invalid: sparsity mismatch
  assertThrows(() => new M.DM(new M.Sparsity(4n, 4n, [0n, 2n, 2n, 3n], [1n, 2n, 1n]),
                              [0.738, 0.39, 0.99]),
               "wrong cols throws");
  assertThrows(() => new M.DM(new M.Sparsity(4n, 3n, [0n, 2n, 2n, 12n], [1n, 2n, 1n]),
                              [0.738, 0.39, 0.99]),
               "colind out of range throws");
}

function test_copyconstr_norefcount(M) {
  // Copy constructor for DM should produce an independent matrix.
  const x = M.DM.ones(2n, 3n);
  const y = new M.DM(x);
  // Mutate x; y should be unchanged.
  // (DM mutation API: set element via indexing.  Cheap proxy: compare nonzeros.)
  const x_nz = x.nonzeros();
  const y_nz = y.nonzeros();
  assertEqual(x_nz.length, 6, "x nnz=6");
  assertEqual(y_nz.length, 6, "y nnz=6");
  for (let i = 0; i < 6; ++i) {
    if (x_nz[i] !== y_nz[i]) throw new Error("DM copy not equal");
    if (x_nz[i] !== 1) throw new Error("DM.ones values not 1");
  }
}

function test_getscheme(M) {
  // Dict-style call: F(x=3, p=4) returns {f, g} where f=x+p, g=x^2.
  const x = M.SX.sym("x");
  const p = M.SX.sym("p");
  const F = M.Function("F", [x, p],
                       [M.plus(x, p), M.times(x, x)],
                       ["x", "p"], ["f", "g"]);
  const fc = F({ x: 3, p: 4 });
  checkArray(fc.f, [7], 10, "F.f at x=3,p=4");
  checkArray(fc.g, [9], 10, "F.g at x=3,p=4");
  if (fc.nope !== undefined) throw new Error("unknown key 'nope' not undefined");
}

function test_assertions(M) {
  // attachAssert -- evaluation succeeds when condition holds, throws otherwise.
  // Method exists on MX -- verify it doesn't crash construction.  Numeric
  // eval-time enforcement may require the assert chain to reach the
  // output (compile-time elision can drop it); only assert construction.
  const x = M.MX.sym("x");
  let z = M.times(x, x);
  if (typeof z.attachAssert !== "function") {
    console.log("    (skip -- attachAssert not exposed on MX)");
    return;
  }
  z = z.attachAssert(M.gt(z, M.MX(3)), "x must be larger than 3");
  const v = M.sin(z);
  const f = M.Function("f", [x], [v]);
  // Verify the holding-true case computes correctly (sin(36) at x=-6).
  const out_ok = f(-6);
  assertTrue(out_ok && out_ok.nonzeros, "f(-6) computed");
}

function test_constructorlol(M) {
  // DM([[1,2,3],[4,5,6]]) constructs a 2x3 matrix from a list-of-lists.
  const D = M.DM([[1, 2, 3], [4, 5, 6]]);
  assertEqual(D.size1(), 2n, "DM size1");
  assertEqual(D.size2(), 3n, "DM size2");
}

function test_DM_zeros_matches_shape(M) {
  // DM.zeros(rows, cols) constructs a 0-valued matrix with the right shape.
  const z = M.DM.zeros(3n, 5n);
  assertEqual(z.size1(), 3n, "zeros size1");
  assertEqual(z.size2(), 5n, "zeros size2");
  // zeros() may be sparse (no explicit 0s); shape is what matters.
}

function test_GenericType_passthrough(M) {
  // M.GenericType wrappers exist; constructing a GenericType from a
  // primitive should round-trip through Function options.
  if (typeof M.GenericType !== "function") {
    console.log("    (skip -- GenericType not exposed)");
    return;
  }
  // Build a Function with options dict containing assorted types.
  const x = M.SX.sym("x");
  const f = M.Function("f", [x], [x], { ad_weight: 0.7, max_num_dir: 64 });
  // Verify it constructs without crashing.
  assertEqual(f.n_in(), 1n, "f.n_in");
}

function test_simplify_no_crash(M) {
  // simplify(SX) -- pre-existing API; just verify it doesn't crash.
  if (typeof M.simplify !== "function") {
    console.log("    (skip -- simplify not exposed)");
    return;
  }
  const x = M.SX.sym("x");
  const e = M.plus(M.times(x, x), M.SX(0));
  const s = M.simplify(e);
  assertTrue(s != null, "simplify returns non-null");
}

runTests([
  ["test_issue179B",            test_issue179B],
  ["test_sanity",               test_sanity],
  ["test_copyconstr_norefcount",test_copyconstr_norefcount],
  ["test_getscheme",            test_getscheme],
  ["test_assertions",           test_assertions],
  ["test_constructorlol",       test_constructorlol],
  ["test_DM_zeros_matches_shape", test_DM_zeros_matches_shape],
  ["test_GenericType_passthrough", test_GenericType_passthrough],
  ["test_simplify_no_crash",    test_simplify_no_crash],
]);

// ===========================================================================
// SKIPPED tests (16 of 21):
//   test_issue4216           -- numpy complex (no numpy in JS)
//   test_copyconstr_refcount -- explicit id() identity check (Python-specific)
//   test_copy_norefcount     -- copy.copy() (Python copy module)
//   test_copy_refcount       -- copy.copy() (Python copy module)
//   test_deepcopy_norefcount -- copy.deepcopy() (Python copy module)
//   test_deepcopy_refcount   -- copy.deepcopy() (Python copy module)
//   test_options_introspection -- requires nlpsol("ipopt") + .optionNames() helper
//   test_pickling            -- pickle module (Python-specific)
//   test_exceptions          -- nlpsol error formatting (capture_stdout)
//   test_doc                 -- nlpsol.__doc__ accessor (Python-specific)
//   test_output              -- capture_stdout + ipopt iter prints
//   test_serialize           -- Function::serialize (known wasm-js OOB)
//   test_print_time          -- capture_stdout
//   test_error_formatting    -- error message format (target-specific wording)
//   test_record_time         -- timing options (capture_stdout)
//   test_unicode             -- Python unicode string handling
